### PR TITLE
chore(types): turn on `strictFunctionTypes` compiler flag

### PR DESCRIPTION
**Pull request by alicewriteswrongs**
This turns the `strictFunctionTypes` compiler flag on and fixes the errors uncovered by doing so.

## Pull request checklist

Please check if your PR fulfills the following requirements:
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been reviewed and added / updated if needed (for bug fixes / features)
- [x] Build (`npm run build`) was run locally and any changes were pushed
- [x] Unit tests (`npm test`) were run locally and passed
- [x] E2E Tests (`npm run test.karma.prod`) were run locally and passed
- [x] Prettier (`npm run prettier`) was run locally and passed

## Pull request type

 

 

Please check the type of change your PR introduces:
- [ ] Bugfix
- [ ] Feature
- [x] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## What is the current behavior?

`strictFunctionTypes` is turned off.

## What is the new behavior?

`strictFunctionTypes` is turned on!

## Does this introduce a breaking change?

- [ ] Yes
- [x] No

## Testing

I ran the tests, made sure everything compiles. I believe these are all only typesystem-level changes.

### CODE REVIEW

File: src/declarations/stencil-private.ts [hunk 1]
```diff
@@ -1145,7 +1145,7 @@ export interface DevServerContext {
   getBuildResults: () => Promise<CompilerBuildResults>;
   getCompilerRequest: (path: string) => Promise<CompilerRequestResponse>;
   isServerListening: boolean;
-  logRequest: (req: { method: string; pathname?: string }, status: number) => void;
+  logRequest: (req: HttpRequest, status: number) => void;
```
Comment (rwaskiewicz) [inline src/declarations/stencil-private.ts:1148]: I may be totally missing something here - I pulled down `main@HEAD` and enabled `strictFunctionTypes` (so any errors you fixed here wouldn't be on my branch and I would see the causers). I don't see a cause/error related to this change though. Can you point me to where/what we needed to update?

**Review by rwaskiewicz — COMMENTED** [review]
One question

File: src/declarations/stencil-private.ts (hunk 1 above)
Comment (alicewriteswrongs) [reply src/declarations/stencil-private.ts:1148]: I see this error:

```
[11:56:24 AM] Starting compilation in watch mode...

src/dev-server/server-context.ts:126:5 - error TS2322: Type '(req: d.HttpRequest, status: number) => void' is not assignable to type '(req: { method: string; pathname?: string; }, status: number) => void'.
  Types of parameters 'req' and 'req' are incompatible.
    Type '{ method: string; pathname?: string; }' is missing the following properties from type 'HttpRequest': acceptHeader, url, searchParams

126     logRequest,
        ~~~~~~~~~~

  src/declarations/stencil-private.ts:1149:5
    1149     logRequest: (req: { method: string; pathname?: string }, status: number) => void;
             ~~~~~~~~~~
    The expected type comes from property 'logRequest' which is declared here on type 'DevServerContext'

[11:56:30 AM] Found 1 error. Watching for file changes.
```

Comment (alicewriteswrongs) [reply src/declarations/stencil-private.ts:1148]: I get that error both on `main` if I do

```sh
npx tsc --noEmit --strictFunctionTypes
```

and if I undo that single line on `ap/strict-function-types`

Comment (rwaskiewicz) [reply src/declarations/stencil-private.ts:1148]: Yuppppp classic PEBKAC. Thanks!

**Review by rwaskiewicz — APPROVED** [review]

**Review by tanner-reits — APPROVED** [review]

**Pull request merged**